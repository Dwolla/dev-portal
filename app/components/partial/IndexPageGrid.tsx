import React from "react";
import styled from "@emotion/styled";
import { useRouter } from "next/router";
import Banner from "../base/Banner";
import FilterTabs from "../base/FilterTabs";
import CardGrid from "../base/CardGrid";
import Card, { LanguageProp } from "../base/Card";
import { slugify } from "../../modules/helpers";
import { contentIcons } from "../../modules/images.import";

const CardGridWrap = styled.div`
  margin-bottom: 20px;
`;

export type CardType = {
  topic: string;
  description: any;
  icon?: string;
  category?: string;
  href: string;
  external?: boolean;
  languages?: Array<LanguageProp>;
};
type Props = {
  unfilteredCards: Array<CardType>;
  bannerProps: {
    icon: string;
    topic: string;
    description: string;
  };
};

function IndexPageGrid({ unfilteredCards, bannerProps }: Props) {
  const router = useRouter();
  const filters: any = [
    { value: null, label: "All" },
    ...new Set(
      unfilteredCards
        .filter((card) => !!card.category)
        .map((card) => card.category)
    ),
  ];
  const filteredCards = unfilteredCards.filter(
    (card) =>
      !router.query.filter || router.query.filter === slugify(card.category)
  );

  const setFilter = (selectedFilter) => {
    if (selectedFilter.value) {
      router.push({
        pathname: router.pathname,
        query: { filter: selectedFilter.value },
      });
    } else {
      router.push(router.pathname);
    }
  };

  return (
    <>
      <Banner
        {...bannerProps}
        filterTabs={
          filters.length > 1 ? (
            <FilterTabs
              tabs={filters}
              filter={
                router.query.filter ? router.query.filter.toString() : null
              }
              setFilter={setFilter}
            />
          ) : null
        }
      />

      <CardGridWrap>
        <CardGrid>
          {filteredCards.map((card) => (
            <Card
              {...card}
              badge={card.category}
              icon={card.icon ? contentIcons[card.icon].default : null}
              links={[
                {
                  href: card.href,
                  text: "GitHub Repository",
                  external: card.external,
                },
              ]}
              key={`${card.topic}-${card.href}`}
              languages={card.languages}
            />
          ))}
        </CardGrid>
      </CardGridWrap>
    </>
  );
}

export default IndexPageGrid;
