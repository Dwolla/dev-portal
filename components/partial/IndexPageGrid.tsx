import React from "react";
import styled from "@emotion/styled";
import { useRouter } from "next/router";
import Banner from "../base/Banner";
import FilterTabs from "../base/FilterTabs";
import CardGrid from "../base/CardGrid";
import Card from "../base/Card";
import { slugify } from "../../modules/helpers";
import { contentIcons } from "../../modules/images.import";
import { breakDown } from "../breakpoints";

const BannerWrap = styled.div`
  margin: -20px -40px 40px -40px;

  @media (${breakDown("sm")}) {
    margin: -20px -20px 20px -20px;
  }
`;

const CardGridWrap = styled.div`
  margin-bottom: 20px;
`;

type CardType = {
  topic: string;
  description: string;
  icon: string;
  category?: string;
  href: string;
};
type Props = {
  unfilteredCards: Array<CardType>;
  bannerProps: {
    icon: string;
    topic: string;
    description: string;
  };
};

const IndexPageGrid = ({ unfilteredCards, bannerProps }: Props) => {
  const router = useRouter();
  const filters: any = [
    { value: null, label: "All" },
    ...new Set(
      unfilteredCards
        .filter((card) => !!card.category)
        .map((card) => card.category)
    ),
  ];
  const filteredCards = unfilteredCards.filter(
    (card) =>
      !router.query.filter || router.query.filter === slugify(card.category)
  );

  const setFilter = (selectedFilter) => {
    if (selectedFilter.value) {
      router.push({
        pathname: router.pathname,
        query: { filter: selectedFilter.value },
      });
    } else {
      router.push(router.pathname);
    }
  };

  return (
    <>
      <BannerWrap className="mdx-style-override">
        <Banner
          {...bannerProps}
          filterTabs={
            filters.length > 1 ? (
              <FilterTabs
                tabs={filters}
                filter={
                  router.query.filter ? router.query.filter.toString() : null
                }
                setFilter={setFilter}
              />
            ) : null
          }
        />
      </BannerWrap>

      <CardGridWrap>
        <CardGrid>
          {filteredCards.map((card) => (
            <Card
              {...card}
              badge={card.category}
              icon={contentIcons[card.icon].default}
              link={{ href: card.href }}
              key={`${card.topic}-${card.href}`}
            />
          ))}
        </CardGrid>
      </CardGridWrap>
    </>
  );
};

export default IndexPageGrid;
