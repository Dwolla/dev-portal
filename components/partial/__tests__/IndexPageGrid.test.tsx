import renderer from "react-test-renderer";
import IndexPageGrid from "../IndexPageGrid";

const mockPath = "/docs/guide";
jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: mockPath,
      pathname: mockPath,
      query: "",
      asPath: mockPath,
    };
  },
}));
jest.mock("../../../modules/images.import");

test("IndexPageGrid", () => {
  const bannerProps = {
    icon: "guides-icon.svg",
    topic: "Guides",
    description:
      "Step-by-step instructions to get you set up in the Dwolla API",
  };
  const guideCards = [
    {
      topic: "Send Money to your Users",
      description:
        "Send ACH transfers. Choose custom (Access API) onboarding or Transfer (OAuth) account creation.",
      icon: "guides-icon.svg",
      category: "Get Building",
      href: "/guides/send-money",
    },
    {
      topic: "Receive Money from your Users",
      description:
        "Accept ACH transfers. Covers both Access API and Transfer integrations.",
      icon: "guides-icon.svg",
      category: "Get Building",
      href: "/guides/receive-money",
    },
    {
      topic: "Transfer Money Between Users",
      description:
        "Facilitate ACH transfers between two distinct parties, e.g. for marketplace applications that connect buyers with sellers for bank to bank payments. Features Access API integration.",
      icon: "guides-icon.svg",
      category: "Get Building",
      href: "/guides/transfer-money-between-users/obtain-access-token",
    },
  ];

  const tree = renderer
    .create(
      <IndexPageGrid unfilteredCards={guideCards} bannerProps={bannerProps} />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
