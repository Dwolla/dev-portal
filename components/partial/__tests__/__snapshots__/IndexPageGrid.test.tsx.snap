// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexPageGrid 1`] = `
Array [
  <div
    className="css-q48buj-BannerWrap e10u28530"
  >
    <div
      className="filters css-1w3gvwd-StyledBanner eyaavg40"
    >
      <img
        alt=""
        className="css-wrcaoo-StyledIcon eyaavg41"
        src="guides-icon.svg"
      />
      <div
        className="css-1ctbztb-StyledTopic eyaavg42"
      >
        Guides
      </div>
      <div
        className="filters css-qmg3be-StyledDescription eyaavg43"
      >
        Step-by-step instructions to get you set up in the Dwolla API
      </div>
      <div
        className="css-5t23wi-StyledFilter eyaavg44"
      >
        <div>
          <div
            className="css-18c9go6-TabsContainerStyle efq5uj30"
          >
            <div
              className="tab is-active css-rdl1v4-TabStyle efq5uj31"
              onClick={[Function]}
            >
              All
            </div>
            <div
              className="tab css-rdl1v4-TabStyle efq5uj31"
              onClick={[Function]}
            >
              Get Building
            </div>
          </div>
          <div
            className="css-n6mzzn-SelectWrapper efq5uj32"
          >
            <div
              className=" css-2b097c-container"
              onKeyDown={[Function]}
            >
              <div
                className=" css-1xxipdn-control"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <div
                  className=" css-86rg1c-ValueContainer"
                >
                  <div
                    className=" css-1me8bqy-singleValue"
                  >
                    All
                  </div>
                  <input
                    aria-autocomplete="list"
                    className="css-62g3xt-dummyInput"
                    disabled={false}
                    id="react-select-2-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    readOnly={true}
                    tabIndex="0"
                    value=""
                  />
                </div>
                <div
                  className=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    className=" css-43ykx9-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    className=" css-unjupu-indicatorContainer"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <img
                      alt=""
                      className=" css-vfym8x-StyledDefaultDropdownIcon e1oqaitj0"
                      src="test-image-stub"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="css-1ujvi4q-CardGridWrap e10u28531"
  >
    <div
      className="css-1nr538h-StyledGrid e1k3ap260"
    >
      <div
        className="css-1opn9ux-CardWrapper e1k3ap261"
      >
        <a
          className="css-kunpmx-StyledLink eguc8hf1"
          href="/guides/send-money"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <div
            className=" css-1mrozmg-CardStyle eguc8hf0"
          >
            <img
              alt=""
              className=" css-142sfdf-StyledIcon eguc8hf2"
              src="/assets/images/content-images/content-icons/guides-icon.svg"
            />
            <div
              className="css-vz2qld-BadgeStyle eguc8hf3"
            >
              <div
                className="css-1bdkrme-StyledBadge e1w3mjmy0"
              >
                Get Building
              </div>
            </div>
            <div
              className=" css-bgc4kb-TopicStyle eguc8hf5"
            >
              Send Money to your Users
            </div>
            <div
              className=" css-mk6kj1-DescriptionStyle eguc8hf6"
            >
              Send ACH transfers. Choose custom (Access API) onboarding or Transfer (OAuth) account creation.
            </div>
          </div>
        </a>
      </div>
      <div
        className="css-1opn9ux-CardWrapper e1k3ap261"
      >
        <a
          className="css-kunpmx-StyledLink eguc8hf1"
          href="/guides/receive-money"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <div
            className=" css-1mrozmg-CardStyle eguc8hf0"
          >
            <img
              alt=""
              className=" css-142sfdf-StyledIcon eguc8hf2"
              src="/assets/images/content-images/content-icons/guides-icon.svg"
            />
            <div
              className="css-vz2qld-BadgeStyle eguc8hf3"
            >
              <div
                className="css-1bdkrme-StyledBadge e1w3mjmy0"
              >
                Get Building
              </div>
            </div>
            <div
              className=" css-bgc4kb-TopicStyle eguc8hf5"
            >
              Receive Money from your Users
            </div>
            <div
              className=" css-mk6kj1-DescriptionStyle eguc8hf6"
            >
              Accept ACH transfers. Covers both Access API and Transfer integrations.
            </div>
          </div>
        </a>
      </div>
      <div
        className="css-1opn9ux-CardWrapper e1k3ap261"
      >
        <a
          className="css-kunpmx-StyledLink eguc8hf1"
          href="/guides/transfer-money-between-users/obtain-access-token"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <div
            className=" css-1mrozmg-CardStyle eguc8hf0"
          >
            <img
              alt=""
              className=" css-142sfdf-StyledIcon eguc8hf2"
              src="/assets/images/content-images/content-icons/guides-icon.svg"
            />
            <div
              className="css-vz2qld-BadgeStyle eguc8hf3"
            >
              <div
                className="css-1bdkrme-StyledBadge e1w3mjmy0"
              >
                Get Building
              </div>
            </div>
            <div
              className=" css-bgc4kb-TopicStyle eguc8hf5"
            >
              Transfer Money Between Users
            </div>
            <div
              className=" css-mk6kj1-DescriptionStyle eguc8hf6"
            >
              Facilitate ACH transfers between two distinct parties, e.g. for marketplace applications that connect buyers with sellers for bank to bank payments. Features Access API integration.
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>,
]
`;
